wikipedia_cleanup/data_filter.py

```diff
@@ -0,0 +1,266 @@
+        self._filter_stats.input_num_changes = len(changes)
+        filtered_changes = []
+        start_idx = 0
+        for end_idx in range(len(changes)):
+            if (
+                (
+                    changes[start_idx].value_valid_from.date()
+                    != changes[end_idx].value_valid_from.date()
+                )
+                or changes[start_idx].infobox_key != changes[end_idx].infobox_key
+                or changes[start_idx].property_name != changes[end_idx].property_name
+            ):
+                filtered_changes.extend(
+                    self._filter_for_property(changes[start_idx:end_idx])
+                )
+                start_idx = end_idx
+        self._filter_stats.output_num_changes = len(filtered_changes)
+        return filtered_changes
+
+    def _filter_for_property(self, changes: List[InfoboxChange]) -> List[InfoboxChange]:
+        if len(changes) == 1:
+            return [changes[0]]
+        values_to_occurrences: Dict[Optional[str], int] = {}
+        for change in changes:
+            if change.current_value in values_to_occurrences.keys():
+                values_to_occurrences[change.current_value] += 1
+            else:
+                values_to_occurrences[change.current_value] = 1
+        max_occurrence = max(
+            values_to_occurrences.items(), key=lambda val_occ: val_occ[1]
+        )[1]
+        representative_change = deepcopy(
+            next(
+                filter(
+                    lambda change: values_to_occurrences[change.current_value]
+                    >= max_occurrence,
+                    reversed(changes),
+                )
+            )
+        )
+        representative_change.value_valid_to = deepcopy(changes[-1].value_valid_to)
+        representative_change.value_valid_from = deepcopy(changes[0].value_valid_from)
+        representative_change.num_changes = len(changes)
+        return [representative_change]
+
+
+class DataFilterBotReverts(AbstractDataFilter):
+    def _filter_for_property(self, changes: List[InfoboxChange]) -> List[InfoboxChange]:
+        def change_pair_is_bot_revert(
+            reverted_change: InfoboxChange, bot_revert_change: InfoboxChange
+        ) -> bool:
+            return (
+                reverted_change.current_value == bot_revert_change.previous_value
+                and reverted_change.previous_value == bot_revert_change.current_value
+                and reverted_change.value_valid_from
+                == bot_revert_change.value_valid_from
+                and reverted_change.value_valid_to == reverted_change.value_valid_to
+                and reverted_change.value_valid_to == bot_revert_change.value_valid_from
+            )
+
+        def change_might_be_reverted(change: InfoboxChange) -> bool:
+            return change.value_valid_from == change.value_valid_to
+
+        filtered_change_indices = []
+        for idx in range(len(changes)):
```

Comment:
Change to while look, always look forward to simplify code, directly append changes to filtered_changes.